.github/actions/setup/action.yml

```diff
@@ -0,0 +1,35 @@
+name: "Setup & build"
+description: "Setup node, install dependencies & build"
+runs:
+  using: composite
+  steps:
+    - uses: actions/setup-node@v2
+      with:
+        node-version-file: '.nvmrc'
+    - uses: actions/cache@v2
+      id: build-cache
+      with:
+        path: |
+          packages/bueno/dist
+          packages/headless/dist
+          packages/auth/dist
+          packages/atomic/dist
+          packages/atomic/docs
+          packages/atomic/headless
+          packages/atomic/loader
+          packages/atomic/storybook-static
+          packages/atomic/www
+        key: ${{ github.sha }}
```

Comment:
I chose the commit SHA as the key for the cache, meaning that we build at-least once per commit.

An alternative would be to try to calculate a hash of the source files and any other file that could potentially affect the build. In this case, we could also divide the cache per-package and build each package individually if their cache isn't hit.

This would likely save a substantial amount of time, but I'm worried that adding new packages would become very difficult and I'm worried that we may forget to include some files in the hash that affect the build (e.g., rollout or esbuild configs).

Reply:
Some thoughts in no particular order:

- Since we are specifying paths in the action, if we rename folders, add new folders, or add new packages, we would need to maintain this point. Do we need to specify exact dirs, or could we just list the package dirs? `atomic`, `auth`, `bueno` etc.

- If the cache is on the commit SHA, then the benefit is limited to when a step on the build fails (e.g. e2e tests). This seems reasonable since this is the scenario where one is retrying. As you and Louis said, using the files as the source of truth would increase the scope of the benefit.

- I wonder if esbuild will reduce the need for a cache? Caches are helpful when a step is slow. If the step is less slow though, then the cache decreases in usefulness. No cache -> less config (first bullet) or the chance of a file not refreshing. Perhaps we are optimizing unnecessarily?

Reply:
> Since we are specifying paths in the action, if we rename folders, add new folders, or add new packages, we would need to maintain this point. Do we need to specify exact dirs, or could we just list the package dirs? `atomic`, `auth`, `bueno` etc.

I could keep the whole dir and exclude `node_modules`. It just felt a bit weird to cache everything, but logically I don't see much of an issue with it.

> I wonder if esbuild will reduce the need for a cache? Caches are helpful when a step is slow. If the step is less slow though, then the cache decreases in usefulness. No cache -> less config (first bullet) or the chance of a file not refreshing. Perhaps we are optimizing unnecessarily?

Perhaps. However, even without the cache, having a setup action reduces the length of the workflow.

Reply:
update: turns out that trying to save everything and excluding node_modules & src creates a cache of about 427MB. Comparatively, if we're more specific, it's 15MB. Given that there's a limit on how much cache we can have, I will stay with the more specific solution.